src/PiWeb.Volume/Block/BlockVolumeDecoder.cs

```diff
@@ -109,18 +132,27 @@ private static void DecodeLayer(
 				//2. ZigZag
 				ZigZag.Reverse( doubleBuffer1, doubleBuffer2 );
 
+				var nonEmptyVectors = FindNonEmptyVectors( doubleBuffer2 );
+
 				//3. Quantization
 				Quantization.Apply( quantization.AsSpan(), doubleBuffer2 );
 
 				//4. Cosine transform
-				DiscreteCosineTransform.Transform( doubleBuffer2, doubleBuffer1, true );
+				DiscreteCosineTransform.Transform( doubleBuffer2, doubleBuffer1, true, nonEmptyVectors );
 
 				//5. Discretization
 				for( var i = 0; i < BlockVolume.N3; i++ )
 					byteBuffer[ i ] = (byte)Math.Clamp( doubleBuffer1[ i ], byte.MinValue, byte.MaxValue );
 
 				blockAction( byteBuffer, blockIndex );
 
+#if DEBUG
+		}
+
+		ArrayPool<double>.Shared.Return( buffers.Item1 );
```

Comment:
You might also create a data structure holding these 3 buffers - the code would be easier readable and you can move the Rent/Return logic into a method of that structure.